### ISSUE TITLE
Frame constructor coerces mixed inputs of floats and strings to strings

### ISSUE THREAD

**Issue by ForeverWintr** [InvestmentSystems/static-frame#122]
Attempting to construct a frame with two columns, one a string and another a float, instead I get two string columns!

```python
f = sf.Frame(list(zip(['a' for x in range(5)], [1.0 for x in range(5)])), columns=['a','b'])
f.dtypes
<Series>
<Index>
a        <U3
b        <U3
<<U1>    <object>
```

**Comment by flexatone** [comment]
Thanks for isolating this issue.

Note that this is the same behavior as the NP array constructor, which is what is being called:

```python
>>> np.array(list(zip(['a' for x in range(5)], [1.0 for x in range(5)])))
array([['a', '1.0'],
       ['a', '1.0'],
       ['a', '1.0'],
       ['a', '1.0'],
       ['a', '1.0']], dtype='<U3')
```
A fine alternative is to use `Frame.from_records`, which does its own type induction:

```python
>>> sf.Frame.from_records(zip(['a' for x in range(5)], [1.0 for x in range(5)]))
<Frame>
<Index> 0     1         <int64>
<Index>
0       a     1.0
1       a     1.0
2       a     1.0
3       a     1.0
4       a     1.0
<int64> <<U1> <float64>
```
Which leaves the question: what should the `Frame` default constructor do?

Presently, The `Frame` constructor takes `TypeBlocks`, `np.array`, and strings or objects that do not have `__len__` (single elements). These are all sensible.

For objects that have `__len__`, we simply call `np.array()`, which nicely handles things that realize 1D arrays. We could intervene here and identify objects that are iterables that contain iterables, and then call `Frame.from_records`. But having the default initializer call a specialized initializer seems undesirable.

We could raise on such cases, and redirect the user to use `Frame.from_records`. But the convention of giving a list of lists to a `Frame` constructor is pretty common.

What do you suggest?

**Comment by ForeverWintr** [comment]
I am a big fan of the 'smart constructors; stupid initializers' approach, so I like the idea of keeping `Frame`'s initializer simple. However, I don't think comparing to numpy is the right thing to do in this case, because in your numpy example above we are expecting to create a single numpy array. When creating a multi column Frame, I think the user *expects* to be creating multiple Series objects (which is what a  `Frame` appears to be). The fact that SF stores multiple columns in a single frame internally is and should remain opaque to the user. 

I think numpy's behavior is less than ideal in this case too, but that's a separate issue.

I can think of two options for [`Frame`'s `__init__`](https://github.com/InvestmentSystems/static-frame/blob/master/static_frame/core/frame.py#L1647):

1. (Friendly) Do a single pass through `data` to check for mixed dtypes. If there are mixed dtypes, fail, directing users to a more specialized interface. We might also require that `__init__`  be called with a `dtypes` argument, but I see that `__init__` doesn't take a dtype argument so I'm not sure if that requirement would be difficult to add. 

2. (Less friendly, and my preference) Forbid untyped data entirely in `__init__`. That is, instead of taking `tp.Iterable[tp.Any]` as `data`, Take only an array/iterable of arrays (or other internal SF data structure, as appropriate). If `__init__` is called with untyped data, fail, directing users to a specialized constructor that does type inspection.

**Comment by flexatone** [comment]
Thanks for these thoughts.

I agree that the expectation of a NP array constructor is different than an SF `Frame` constructor.

Regarding (1): this is expensive, and that is already implemented to some degree in `Frame.from_records`, so this is not appealing to me either. Adding `dtypes` is also undesirable, as it would suggest that we will retype already typed data, and this does not seem like the place to do that. 

Regarding (2): this is generally my current thinking as well, though your approach is a bit more restrictive than my present thoughts on this. I agree that anything that looks like it should be a 2D table, better handled by `Frame.from_records`, should raise, and the user should be directed to use `Frame.from_records`. But there are two other untyped cases that I feel the default constructor might handle:

a: elements, i.e., single items like one number, one string, NaN, None, etc. A specialized constructor for this case seems a little too heavy, and the type induction on this is obvious, fast, and foolproof.

b: 1D non-array iterables that produce a `Frame` with a single column. It is hard to imagine what a clear name for this constructor would be (ideas?), and it is a case, like `Series` default constructor, for which we already have the tools to do decent type induction.

**Comment by ForeverWintr** [comment]
Now I want `Frame.from_element` and `Frame.from_column`!

**Comment by flexatone** [comment]
I am swayed by this consideration: if the constructor needs dtype(s) to be specified, it should be specialized constructor. This suggest adding the following.

`Frame.from_element(element, dtype)`
`Frame.from_record(record, dtype)`

For parallelism, we should add to `Series` the following:

`Series.from_element(element, dtype)`

For both `Series` and `Frame`, we will remove support for these types in the default constructor.

**Comment by ForeverWintr** [comment]
I think "if the constructor needs dtype(s) to be specified, it should be specialized constructor" is a great heuristic!

**Comment by flexatone** [comment]
To be released in 0.5.12 are the following new constructors:

`Frame.from_element`
`Frame.from_series`
`Frame.from_elements` (described previously as `from_record`,  but  that name was rejected as it sounds too much like providing a row rather than a column.)
`Series.from_element`

With the introduction of these constructors (as well as the existing `Frame.from_records`), creating a `Frame` from single elements or untyped iterables will raise a deprecation warning, and will ultimately raise an exception by 0.6. Similarly, creating a `Series` from a single element is deprecated and will raise an exception in 0.6.

Returning to the example that we started with, we can see the warning and the alternative approach, below:

```python
In [2]: f = sf.Frame(list(zip(['a' for x in range(5)], [1.0 for x in range(5)])), columns=['a','b'])                                                                    
/home/ariza/src/static-frame/static_frame/core/frame.py:1896: UserWarning: The Frame initializer no longer supports creation from untyped iterables; use Frame.from_records to create a Frame from an iterable of iterables; use Frame.from_elements to create a Frame from an iterable.
  deprecated('The Frame initializer no longer supports creation from untyped iterables; use Frame.from_records to create a Frame from an iterable of iterables; use Frame.from_elements to create a Frame from an iterable.')

In [3]: f = sf.Frame.from_records(list(zip(['a' for x in range(5)], [1.0 for x in range(5)])), columns=['a','b'])                                                       

In [4]: f                                                                                                                                                               
Out[4]: 
<Frame>
<Index> a     b         <<U1>
<Index>
0       a     1.0
1       a     1.0
2       a     1.0
3       a     1.0
4       a     1.0
<int64> <<U1> <float64>
```
